### PR TITLE
Crop API add aspect ratio field for future use by UI imageWidgetCrop …

### PR DESCRIPTION
**Pull request by woprrr**
This is final change for render crop integration 100% compatible with CROP API.

I ve add aspec_ratio field into the form crop_type entity for set an aspect ratio for crop_types. If the user need to have an free aspect ratio, is possible to stay empty to have a free-zone.

I ve add two helpers for get the crop_types ans crop_type_names like node ;).

Thanks @slashrsm for your support :) :+1:

### CODE REVIEW

**Comment by woprrr** [comment]
I ve delete the deprecated function introduced. I ve found best solution to get the list intead of my widget class. Juste get_names is required and very usefull :+1:

**Comment by woprrr** [comment]
@slashrsm I ve a question for u, i try to add a test in new aspect_ratio field in method (testCropTypeSave()) but i can't use random because i need inter an format like "W:H" eg : 16:9 and not a string (i ve added a validator to check it if user have set an custom aspect_ratio).

Edit : I ve think i can test it instead of ImageWidgetCrop because i use it in widget.

**Comment by woprrr** [comment]
In my way i Fix deprecated use class in Tests (@see refs #2580001 Investigate why tests fail on HHVM and fix them).

File: crop.module [hunk 1]
```diff
@@ -88,3 +88,17 @@ function crop_form_media_bundle_form_alter(&$form, FormStateInterface $form_stat
...
+
+/**
+ * Returns a list of available crop type names.
+ *
+ * This list can include types that are queued for addition or deletion.
+ *
+ * @return string[]
+ *   An array of crop type labels, keyed by the crop type name.
+ */
+function crop_type_get_names() {
```
Comment (slashrsm) [inline crop.module]: It feels that having this as a static function on crop type class would be better.

File: src/Form/CropTypeForm.php [hunk 2]
```diff
@@ -55,6 +55,13 @@ public function form(array $form, FormStateInterface $form_state) {
       '#description' => t('Describe this crop type.'),
     ];
 
+    $form['aspect_ratio'] = [
+      '#title' => t('Aspect Ratio'),
+      '#type' => 'textfield',
+      '#default_value' => $type->aspect_ratio,
+      '#description' => t('Set an aspect ratio eg: <b>16:9</b>, if you set this at (0:0) or empty the crop zone is free.'),
```
Comment (slashrsm) [inline src/Form/CropTypeForm.php]: Maybe something like "Leave this empty for arbitrary aspect ratio."

File: src/Form/CropTypeForm.php [hunk 3]
```diff
@@ -71,14 +78,21 @@ protected function actions(array $form, FormStateInterface $form_state) {
...
+    $aspect_ratio = $form_state->getValue('aspect_ratio');
+
     // '0' is invalid, since elsewhere we check it using empty().
     if ($id == '0') {
       $form_state->setErrorByName('type', $this->t("Invalid machine-readable name. Enter a name other than %invalid.", array('%invalid' => $id)));
     }
+
+    // If aspect ratio is set verify format.
+    if (!empty($aspect_ratio) && !preg_match("#^[0-9:]+$#", $aspect_ratio)) {
+      $form_state->setErrorByName('aspect_ratio', $this->t("Invalid format of aspect ratio. Enter a ratio in the format %format.", array('%format' => 'W:H')));
```
Comment (slashrsm) [inline src/Form/CropTypeForm.php]: Could we convert this into a constraint?

Comment (woprrr) [inline src/Form/CropTypeForm.php]: Yes i think only in this case (if user have set an aspect ratio) we need to verify if this format is really an ratio type W:H to use it in UI. What would be your idea about this?

**Comment by woprrr** [comment]
Thanks @slashrsm to your return :) i ve fix 2 points of 3. The last i am not sure to understand what do you mean and have add a comment to explain my vision ;).

**Comment by woprrr** [comment]
Ow ! I did not know in drupal 8 have a sf contraints !!!! I is i need ! But i need a create a new plugin validator to implement my contraint but i did not say to link my contraint to the form validator you have an example please @slashrsm ?

**Comment by slashrsm** [comment]
drupal-media/media_entity_twitter#9

**Comment by woprrr** [comment]
I ve try to use a constraint but i don't understand were implement it in cropType form :( help @slashrsm

**Comment by vpeltot** [comment]
**Constraint Validation :**
I've added a new Constraint to validate the CropType entity (create / update) and I use this constraint in validateForm.
@slashrsm : Is this ok for you ?

**Other :**
I've added a condition to disable the machine name field on cropType update form.
A beautiful "The website encountered an unexpected error. Please try again later." was generated when I've tried to change the machine name.

**Comment by woprrr** [comment]
Thanks lot @vpeltot :dancers: I've add a small fixes of coding standards. I ve tested it and all is Okay for me seems work a charme. What do you think @slashrsm ?

File: src/Entity/CropType.php [hunk 4]
```diff
@@ -61,10 +66,48 @@ class CropType extends ConfigEntityBundleBase implements CropTypeInterface {
...
+   */
+  public function getIterator() {
+    return new \ArrayIterator();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getCropTypeNames() {
+    return array_map(function ($bundle_info) {
```
Comment (slashrsm) [inline src/Entity/CropType.php]: I am not sure if there is any coding standard related to this, but I prefer something like:

```
return array_map(
  function ($bundle_info) { return $bundle_info['label'];},
  \Drupal::entityManager()->getBundleInfo('crop')
);
```

It seems more readable to me.

File: src/Plugin/Validation/Constraint/CropTypeValidationConstraintValidator.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    $id = trim($value->id());
+    // '0' is invalid, since elsewhere we check it using empty().
+    if ($id == '0') {
+      $this->context->buildViolation('Invalid machine-readable name. Enter a name other than "0"')
+        ->atPath('id')
+        ->addViolation();
+    }
+
+    $aspect_ratio = $value->getAspectRatio();
+    if (!empty($aspect_ratio) && !preg_match($value::VALIDATION_REGEXP, $aspect_ratio)) {
```
Comment (slashrsm) [inline src/Plugin/Validation/Constraint/CropTypeValidationConstraintValidator.php]: Would be great to separate this in two constraints (machine name and aspect ratio). That would also fix better error messages part.

File: src/Plugin/Validation/Constraint/CropTypeValidationConstraint.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * )
+ */
+class CropTypeValidationConstraint extends Constraint {
+
+  /**
+   * The default violation message.
+   *
+   * @var string
+   */
+  public $message = 'Validation Error for the crop type.';
```
Comment (slashrsm) [inline src/Plugin/Validation/Constraint/CropTypeValidationConstraint.php]: Would be great to have error messages that actually tell what the problem is.

**Comment by slashrsm** [comment]
This looks great. Thank you both for your work! Let's fix this last few things and it should be ready to merge.

**Comment by vpeltot** [comment]
I corrected this last few things, everything should be ok now.
@slashrsm ready to merge ?

File: src/Plugin/Validation/Constraint/CropTypeAspectRatioValidationConstraint.php [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * )
+ */
+class CropTypeAspectRatioValidationConstraint extends Constraint {
+
+  /**
+   * The default violation message.
+   *
+   * @var string
+   */
+  public $message = 'Invalid format of aspect ratio. Enter a ratio in format H:W.';
```
Comment (slashrsm) [inline src/Plugin/Validation/Constraint/CropTypeAspectRatioValidationConstraint.php]: Maybe something like:

"Invalid aspect ratio format. Should be defined in H:W form."

File: src/Plugin/Validation/Constraint/CropTypeAspectRatioValidationConstraint.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+}
```
Comment (slashrsm) [inline src/Plugin/Validation/Constraint/CropTypeAspectRatioValidationConstraint.php]: Missing newline.

File: src/Plugin/Validation/Constraint/CropTypeMachineNameValidationConstraint.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * )
+ */
+class CropTypeMachineNameValidationConstraint extends Constraint {
+
+  /**
+   * The default violation message.
+   *
+   * @var string
+   */
+  public $message = 'Invalid machine-readable name. Enter a name other than "0".';
```
Comment (slashrsm) [inline src/Plugin/Validation/Constraint/CropTypeMachineNameValidationConstraint.php]: Second sentence seems strange....

File: src/Plugin/Validation/Constraint/CropTypeMachineNameValidationConstraint.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+}
```
Comment (slashrsm) [inline src/Plugin/Validation/Constraint/CropTypeMachineNameValidationConstraint.php]: Missing newline.

**Comment by woprrr** [comment]
Now @slashrsm ready to merge i think :)

**Comment by slashrsm** [comment]
Merged. Thank you both!

**Pull request merged**